<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>FinisherSC by kakitone</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-dark.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js"></script>
    <script src="javascripts/main.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>
  <body>

      <header>
        <h1>FinisherSC</h1>
        <p>A repeat-aware tool for upgrading de-novo assembly using long reads</p>
      </header>

      <div id="banner">
        <span id="logo"></span>

        <a href="https://github.com/kakitone/finishingTool" class="button fork"><strong>View On GitHub</strong></a>
        <div class="downloads">
          <span>Downloads:</span>
          <ul>
            <li><a href="https://github.com/kakitone/finishingTool/zipball/master" class="button">ZIP</a></li>
            <li><a href="https://github.com/kakitone/finishingTool/tarball/master" class="button">TAR</a></li>
          </ul>
        </div>
      </div><!-- end banner -->

    <div class="wrapper">
      <nav>
        <ul></ul>
      </nav>
      <section>
        <h1>
<a id="finishersc" class="anchor" href="#finishersc" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>FinisherSC</h1>

<h2>
<a id="introduction-and-contact" class="anchor" href="#introduction-and-contact" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Introduction and contact</h2>

<p>If you encounter any issue, please feel free to directly contact me at <a href="mailto:kklam@eecs.berkeley.edu">kklam@eecs.berkeley.edu</a>
My name is  Ka-Kit Lam. </p>

<p>Our paper on this tool is published in</p>

<p>Lam, K. K., LaButti, K., Khalak, A., &amp; Tse, D. (2015). FinisherSC: a repeat-aware tool for upgrading de novo assembly using long reads. Bioinformatics, 31(19), 3207-3209.</p>

<p><a href="https://doi.org/10.1093/bioinformatics/btv280">https://doi.org/10.1093/bioinformatics/btv280</a></p>

<p>Moreover, FinisherSC is under MIT License. </p>

<h2>
<a id="news" class="anchor" href="#news" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>News</h2>

<p>If you have a meta-genomics dataset, you may also want to visit our MetaFinisherSC <a href="https://github.com/kakitone/MetaFinisherSC">https://github.com/kakitone/MetaFinisherSC</a></p>

<h2>
<a id="command-to-run-the-tool" class="anchor" href="#command-to-run-the-tool" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Command to run the tool</h2>

<p>Here is the command to run the tool:</p>

<pre><code>python finisherSC.py destinedFolder mummerPath
</code></pre>

<p>If you are running on server computer and would like to use multiple threads, then the following commands can generate 20 threads to run FinisherSC. </p>

<pre><code>python finisherSC.py -par 20 destinedFolder mummerPath
</code></pre>

<p>Sometimes, if the names of raw reads and contigs consists of special characters/formats, FinisherSC/MUMmer may not parse them correctly. In that case, you want to have a quick renaming of the names of contigs/reads in contigs.fasta or raw_reads.fasta using the following command. </p>

<pre><code>    perl -pe 's/&gt;[^\$]*$/"&gt;Seg" . ++$n ."\n"/ge' raw_reads.fasta &gt; newRaw_reads.fasta
    cp newRaw_reads.fasta raw_reads.fasta
    perl -pe 's/&gt;[^\$]*$/"&gt;Seg" . ++$n ."\n"/ge' contigs.fasta &gt; newContigs.fasta
    cp newContigs.fasta contigs.fasta
</code></pre>

<h2>
<a id="input-and-output-files" class="anchor" href="#input-and-output-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input and output files</h2>

<p>Note that it is assumed that you have "contigs.fasta" , "raw_reads.fasta" in the destinedFolder and mummerPath is your path to Mummer. Therefore, please rename the files in the destinedFolder if needed. For "finisherSC.py", the key output file is <strong>improved3.fasta</strong> in destinedFolder. </p>

<h2>
<a id="a-sample-test-run" class="anchor" href="#a-sample-test-run" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>A sample test run</h2>

<p>Here is an example run with the pre-installed data files and softwares(e.g. Mummer, Gepard) [You may go to <a href="https://www.dropbox.com/sh/xjpt8xf5g1xf0ek/bGmZvt9Zfd">https://www.dropbox.com/sh/xjpt8xf5g1xf0ek/bGmZvt9Zfd</a>  to download the the whole package to have a quick test. But we note here that all data and softwares besides FinisherSC have their copyright belonging to their original authors. We provide the sandbox here for users' convenience only. Moreover, all the updates and versioning will be at GitHub page. Users are encouraged to git clone our project for later usage after they successfully go through the example run].</p>

<ol>
<li>You may run the test cases as python finisherSC.py EcoliTestRun/ MUMmer3.23/</li>
<li>After the processing, take the file whose name is <strong>improved3.fasta</strong> in destinedFolder and this is your output file.</li>
<li>If you want to visualize the output against the reference, you can use software like Gepard.</li>
</ol>

<h2>
<a id="options" class="anchor" href="#options" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Options</h2>

<p>a)Fast mode [ -f True ] (e.g. for large and complex genomes , you may want to use this mode to get results faster; However, there will be a bit of computation vs quality trade-off if you want to use this mode; We suggest that you use this mode only when you got stuck in the standard mode): </p>

<pre><code>python finisherSC.py -f True destinedFolder mummerPath
</code></pre>

<p>[It means that you want to use the fast mode ]</p>

<p>b)Parallel mode. [-par numberOfThreads] You can now use -par 20 to run finisherSC on 20 threads. The command is </p>

<pre><code>python finisherSC.py -par 20 destinedFolder mummerPath
</code></pre>

<p>c)Break down large contig file [-l True]. If your contigs.fasta is too big, then you want to use this option to break it down for alignment. </p>

<pre><code>python finisherSC.py -l True destinedFolder mummerPath
</code></pre>

<p>d)Pick from previous jobs [-p pickupFilename] (e.g. you have got improved2.fasta but the cluster node timeout and you want to start from improved2.fasta instead of from scratch)</p>

<pre><code>python finisherSC.py -p improved2.fasta destinedFolder mummerPath  
</code></pre>

<p>e)Mapping between old contigs and new contigs [-o referenceName_QueryName ](e.g. you want to know how the contigs from contigs.fasta are mapped to improved3.fasta )</p>

<pre><code>python finisherSC.py -o contigs.fasta_improved3.fasta destinedFolder mummerPath 
</code></pre>

<p>[It will then output the alignment of improved3.fasta against contigs.fasta. The output will be shown in the terminal and in mappingResults.txt at the destinedFolder]</p>

<p>f)Help [-h]. You can always use -h to get the usage suggestion. </p>

<pre><code>python finisherSC.py -h
</code></pre>

<p>g)There is an experimental improvement based on repeat phasing and the further improved file is improved4.fasta. An illustration of that part is given in Fig. 7 of <a href="http://arxiv.org/abs/1402.6971">http://arxiv.org/abs/1402.6971</a> . To experiment with that, first run finisherSC.py as before. After that, you can issue the following command:</p>

<pre><code>python experimental/xPhaser.py destinedFolder mummerPath
</code></pre>

<p>h) There is also an experimental improvement resolving tandem repeat. The further improved file is tademResolved.fasta. To experiment with that, first run finisherSC.py as before. After that, you can issue the following command:</p>

<pre><code>python experimental/tSolver.py destinedFolder mummerPath
</code></pre>

<h2>
<a id="data-files-for-experiments" class="anchor" href="#data-files-for-experiments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data files for experiments</h2>

<p>We are using the benchmark datasets available online in PacBio DevNet. And we can run FinisherSC to process those data to completion. For (1, 2, 3), we use -par 20 option and run it on a server computer and we get them finished in a couple of hours to a day. For (4), we run without any options on a laptop and get it finished in a couple of minutes. </p>

<ol>
<li>Drosophila : <a href="https://github.com/PacificBiosciences/DevNet/wiki/Drosophila-sequence-and-assembly">https://github.com/PacificBiosciences/DevNet/wiki/Drosophila-sequence-and-assembly</a> </li>
<li>Saccharomyces cerevisiae W303 : <a href="https://github.com/PacificBiosciences/DevNet/wiki/Saccharomyces-cerevisiae-W303-Assembly-Contigs">https://github.com/PacificBiosciences/DevNet/wiki/Saccharomyces-cerevisiae-W303-Assembly-Contigs</a>
</li>
<li>C. elegans : <a href="https://github.com/PacificBiosciences/DevNet/wiki/C.-elegans-data-set">https://github.com/PacificBiosciences/DevNet/wiki/C.-elegans-data-set</a>
</li>
<li>E. coli, M. ruber and P. heparinus : <a href="http://files.pacb.com/software/hgap/index.html">http://files.pacb.com/software/hgap/index.html</a>
</li>
<li>Column (a,b) in the main table of the paper : <a href="http://www.eecs.berkeley.edu/%7Ekakitone/Ped30xData.zip">http://www.eecs.berkeley.edu/~kakitone/Ped30xData.zip</a> , <a href="http://www.eecs.berkeley.edu/%7Ekakitone/Ped50xData.zip">http://www.eecs.berkeley.edu/~kakitone/Ped50xData.zip</a>
</li>
</ol>

<p>The data are supported by the original authors. But we provide an example on how to download and transform data types so it is easier for users to validate FinisherSC on benchmark data . </p>

<ol>
<li>
<p>Download data with links specified in file_list</p>

<pre><code>for f in `cat file_list`; do wget --force-directories $f; done 
</code></pre>
</li>
<li>
<p>Download the DEXTRACTOR</p>

<pre><code>git clone https://github.com/thegenemyers/DEXTRACTOR.git
</code></pre>
</li>
<li>
<p>Transform .bax.h5 files to .fasta file</p>

<pre><code>find . -name '*.bax.h5' | xargs DEXTRACTOR/dextract  &gt; raw_reads.fasta
</code></pre>
</li>
</ol>

<h2>
<a id="remarks" class="anchor" href="#remarks" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Remarks</h2>

<ol>
<li>After passing through the finishingTool, it is advised to use a polishing tool(e.g. Quiver) to polish the improved contigs( i.e. improved3.fasta )</li>
<li>If you have problem running scipy.weave, try to remove ~/.python27_compiled </li>
<li>
<p>Sometimes, if the names of raw reads and contigs consists of special characters/formats, FinisherSC/MUMmer may not parse them correctly. In that case, you want to have a quick renaming of the names of contigs/reads in contigs.fasta or raw_reads.fasta using the following command. </p>

<pre><code>perl -pe 's/&gt;[^\$]*$/"&gt;Seg" . ++$n ."\n"/ge' raw_reads.fasta &gt; newRaw_reads.fasta
cp newRaw_reads.fasta raw_reads.fasta
perl -pe 's/&gt;[^\$]*$/"&gt;Seg" . ++$n ."\n"/ge' contigs.fasta &gt; newContigs.fasta
cp newContigs.fasta contigs.fasta
</code></pre>
</li>
</ol>

<h2>
<a id="annotation-of-finishersc" class="anchor" href="#annotation-of-finishersc" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Annotation of FinisherSC</h2>

<p>Main functional components of FinisherSC : </p>

<blockquote>
<p>finisherSC.py : Starting point of FinisherSC</p>

<p>nonRedundantResolver.py : Filter completely embedded contigs (aka Subroutine 1 in the paper's supplementary material)</p>

<p>overlapResolver.py : Utilize existing overlaps of contigs to resolve repeats and merge contigs (aka Subroutine 2,3 in the paper's supplementary material)</p>

<p>gapFiller.py : Utilize raw reads data to fill gaps(aka Subroutine 4, 5 in the paper's supplementary material)</p>

<p>twoRepeatOneBridgeSolver.py : Resolve repeats with two copies where only one copy is spanned by some reads (aka Subroutine 6,7 in the paper's supplementary material)</p>
</blockquote>

<p>Helper Functions :</p>

<blockquote>
<p>IORobot.py : various functions that support reading and writing of data</p>

<p>alignerRobot.py : library containing ways to parse MUMmer results</p>

<p>houseKeeper.py : various functions to support house keeping of data</p>

<p>graphLib.py : library containing various graph string graph operations </p>

<p>debugging.py : a debugging point for hacking into various functions </p>

<p>unittester.py : integration test package</p>

<p>viewer.py : create dot plots </p>
</blockquote>

<p>An older version of the tool (no longer supported) : </p>

<blockquote>
<p>finisher.py : an independent and functioning(but no longer supported) version based on a greedy algorithm</p>
</blockquote>

<p>Relevant third parties software: </p>

<blockquote>
<p>fasta-splitter.pl : a tool to break FASTA file into smaller trucks in commandline</p>

<p>MUMmer : an alignment tool to perform the mapping of reads and contigs</p>

<p>gepard-1.30 : a good tool to create dotplots</p>
</blockquote>

<p>Experimental folder: </p>

<blockquote>
<p>There are various experimental functions there. The two important ones are tandem repeat resolver(tSolver.py) and approximate repeat phaser(xPhaser.py). </p>
</blockquote>

<h2>
<a id="current-build-status" class="anchor" href="#current-build-status" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Current build status</h2>

<p>It is an indicator on the current built powered by Travis-CI. If you issue a pull request, Travis-CI will evaluate your suggestion by automatically running the code on the default test case. </p>

<p><img src="https://travis-ci.org/kakitone/finishingTool.svg?branch=master" alt="alt text" title="Current build status"></p>
      </section>
      <footer>
        <p>Project maintained by <a href="https://github.com/kakitone">kakitone</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://twitter.com/michigangraham">mattgraham</a></small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-52516611-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
